### ISSUE TITLE
map as an addon?

### ISSUE THREAD

**Issue by domagojk** [gcanti/io-ts#108]
I really like "io-ts" (impressive work @gcanti!), but since version 0.9.0 `map` is no long supported I had to lock my dependency to 0.8.2 :disappointed:.

So, I was just wondering, could `map` be used "as an addon"?

I realize there is problem with serializing it, but I'm interested only in validation part... 
Maybe there could be a way to make it available but with a notice in documentation that "it's not seriazable"?

I don't know if that makes sense?

P.S.

If you're interested in my use case, you can read about it [here](https://medium.com/@domagojk/patterns-for-designing-flexible-architecture-in-node-js-cqrs-es-onion-7eb10bbefe17) (for io-ts part, scroll down to code screenshots).
Here is [a part](https://github.com/domagojk/beenion/blob/a3c392e5e81a28b8a884f1f991da8760156c3fa4/beenion-core/domain/types/journal/events.ts#L46) where `map` is used.

**Comment by gcanti** [comment]
@domagojk Are you interested in a generic `map` (I'm afraid we can't bring it back..) or in your particular use case, i.e. "extracting" the runtime type of a field contained in each member of a union (*)?

(*)

```ts
// io-ts v0.9.3

const pluck = <F extends string, U extends t.UnionType<Array<t.InterfaceType<{ [K in F]: t.Any }, any>>, any>>(
  union: U,
  field: F
): t.Type<any, t.TypeOf<U>[F]> => {
  return t.union(union.types.map(type => type.props[field]))
}

export const UserEvent = t.union([
  t.type({
    type: t.literal('UserCreated'),
    payload: t.type({
      foo: t.string
    })
  }),
  t.type({
    type: t.literal('ReviewInvitationDeclined'),
    payload: t.type({
      bar: t.string
    })
  })
])

// UserEventType: t.Type<any, "UserCreated" | "ReviewInvitationDeclined">
const UserEventType = pluck(UserEvent, 'type')
```

**Comment by domagojk** [comment]
@gcanti Oh.. I didn't know about this.

Thank you, that's all I need here! :relieved: 👏

**Comment by domagojk** [comment]
@gcanti Maybe it would be valuable to include a `helpers.ts` (or similar) file which would include functions like this one?

I'm saying this because I don't think I would come up with this solution (at least not so quickly as you).

Also, I'm not really "advanced FP" developer (I love writing simple pure functions but it's so not natural for me to think in monads, or using either, fold and similar).

However I still love io-ts!

So I think a few functions which would abstract this kind of complexity would be attractive to developers like myself.

Besides `plunk`, in this kind of file, I would also include `brand` (which was featured in README in the past)(https://github.com/domagojk/beenion/blob/master/beenion-core/domain/types/input.ts#L106) and "more imperative" approach to validation [like this](https://github.com/domagojk/beenion/blob/master/beenion-core/domain/validateCommand.ts), so it can be used without `fold` ([screenshot](https://cdn-images-1.medium.com/max/2000/1*v4bMFdtzBBI98bu-LVrHhg.png))

**Comment by sledorze** [comment]
@domagojk you may find io-ts related stuff as part of https://github.com/gcanti/io-ts-types

**Comment by gcanti** [comment]
@domagojk I prefer to keep this library lean, though I just added `pluck` in the readme as a further example of custom combinators, thanks for pointing out.

The `brand` combinator is superseded by `newtype-ts` + `io-ts-types/lib/newtype-ts/fromNewtype`.

`validateCommand`, rather than more imperative, looks unsafe :) I'm not acquainted with the codebase so I'm not sure why in the screenshot `command` is typed `object` instead of `t.TypeOf<typeof AddJournalEditor>` (supposing `AddJournalEditor` a runtime type)

**Comment by domagojk** [comment]
@gcanti My reasoning was that since `command` is the input to the "system" I can't be sure what will actually be passed to it. So at first I've decided to use `any` for its type, but if I forget to validate the command, anything passed to a function defined in a "domain" layer can be correct, and therefore TypeScript won't complain.

But when I use `object` as the input type, TypeScript will raise an error if I don't validate it first (for example [here](https://github.com/domagojk/beenion/blob/master/beenion-core/commandHandlers/article.ts#L193)).

Only after the validation from io-ts, a type is inferred and I can be sure that the command is indeed `ResubmitArticle` which is expected in a [function defined in a domain layer](https://github.com/domagojk/beenion/blob/master/beenion-core/domain/entities/article/methods/resubmit.ts).

However, If you think this is unsafe, could you elaborate or propose a non-imperative alternative? 
I wouldn't be surprised if I'm doing something wrong so for any suggestion I would be most grateful! :)

**Comment by gcanti** [comment]
> command is the input to the "system"

Ah I see, thanks for explaining.

> But when I use object as the input type, TypeScript will raise an error if I don't validate it first (for example here).

That's interesting, there's a [proposal](https://github.com/gcanti/io-ts/issues/53#issuecomment-345864979) to use the following type to represent an unvalidated input

```
type Unknown = {} | null | void
```

> if I'm doing something wrong

No no, nothing wrong. I say unsafe because is throwing, which generally I consider a dangerous behaviour (in every style). I'm just used to return an Option / Either to represent a computation that can fail.

**Comment by domagojk** [comment]
`Unknown` is much more clear than `object`, I will use that instead, thanks!

>  which generally I consider a dangerous behaviour (in every style). I'm just used to return an Option / Either to represent a computation that can fail.

Understood! :)
